R/visualize.R

```diff
@@ -0,0 +1,239 @@
+#'
+#' @param object ([PredictionClassif])\cr ([BenchmarkResult])\cr ([ResampleResult])\cr
+#' The binary class prediction object that will be evaluated. Only one data task is allowed for BenchmarkResult or ResampleResult
+#' * If provided a ([PredictionClassif]). Then the visualization will compare the fairness metrics among the binary level from protected field through bar plots.
+#' * If provided a ([ResampleResult]). Then the visualization will generate the boxplots for fairness metrics, and compare them among the binary level from protected field.
+#' * If provided a ([BenchmarkResult]). Then the visualization will generate the boxplots for fairness metrics, and compare them among both the binary level from protected field and the models implemented.
+#'
+#' @param ...
+#' The arguments to be passed to methods, such as:
+#'
+#' * `fairness_measure` ([Measure])\cr
+#' The fairness measures that will evaluated on object, could be single measure \link{msr} or measures \link{msrs}.  Default measure set to be \code{msr("fairness.acc")}
+#'
+#' * `task` ([TaskClassif])\cr
+#' The data task that contains the protected column, only required when the class of object is ([PredictionClassif])
+#'
+#' @examples
+#' library(mlr3fairness)
+#' library(mlr3learners)
+#' library(mlr3)
+#' library(ggplot2)
+#' library(data.table)
+#'
+#' # Setup the Fairness Measures and tasks
+#' task = tsk("adult_train")
+#' learner = lrn("classif.ranger", predict_type = "prob")
+#' learner$train(task)
+#' predictions = learner$predict(task)
+#' design = benchmark_grid(
+#'   tasks = tsk("adult_train"),
+#'   learners = lrns(c("classif.ranger", "classif.rpart"),
+#'                  predict_type = "prob", predict_sets = c("train", "test")),
+#'   resamplings = rsmps("cv", folds = 3)
+#' )
+#'
+#' bmr = benchmark(design)
+#' fairness_measure = msr("fairness.tpr")
+#' fairness_measures = msrs(c("fairness.tpr", "fairness.fnr"))
+#'
+#' fairness_compare(predictions, fairness_measure, task)
+#' fairness_compare(predictions, fairness_measures, task)
+#' fairness_compare(bmr, fairness_measure)
+#' fairness_compare(bmr, fairness_measures)
+#' @export
+fairness_compare <- function(object, ...){
+  UseMethod("fairness_compare")
+}
+
+#' @export
+fairness_compare.PredictionClassif <- function(object, fairness_measure = msr("fairness.acc"), task, ...){
+  measures = object$score(fairness_measure, task)
+  data <- melt(data.table(names = names(measures), data = measures), id.vars = "names")
+  ggplot(data, aes(x=names, y=value, fill = names)) +
+    geom_bar(stat = "identity",  width=0.4/length(unique(data$names))) +
+    xlab("metrics") +
+    ylab("values") +
+    theme(legend.position = "none") +
+    scale_fill_hue(c=100, l=60)
+}
+
+#' @export
+fairness_compare.BenchmarkResult <- function(object, fairness_measure = msr("fairness.acc"), ...){
+  fairness_name = get_msrs_name(fairness_measure)
```

Comment:
Use `assert_measure(as_measure(fairness_measure))` to convert (if possible) and and assert the class, then `mlr3misc::ids()` to get the name(s).

Reply:
Both assert_measure and as_measure cannot be applied on `msrs()` since it returned a list of msr which functions are not supported, but ids() work as expected so I have changed it.

Reply:
Done

Reply:
If you want to allow multiple measures, I would suggest (1) renaming the parameter to `fairness_measures`, and (2) calling `assert_measures(as_measures(fairness_measures))`.

Reply:
This is what `ResampleResult$score()` calls:

```r
      measures = as_measures(measures, task_type = private$.data$task_type)
      assert_measures(measures, task = self$task, learner = self$learner)
```

By providing the task type in the first line you can query the default measure if none is provided. The additional arguments to `assert_measures()` will ensure that the measure is applicable to the task and learner, respectively.

Reply:
I see, I have changed this to `assert_measures(as_measures(fairness_measures))`. But there is a small problem. Since `fairness_measures` could be one measure `msr()` or a list of measures `msrs()`. And `assert_measures` or `as_measures` could only work for multiple measures, so we need to check the fairness_measure to be one measure or multiple measures.

Reply:
Currently it looks like this 
```
  if("list" %in% class(fairness_measure)) {
    assert_measures(as_measures(fairness_measure))
    fairness_name = ids(fairness_measure)
  } else {
    assert_measure(as_measure(fairness_measure))
    fairness_name = fairness_measure$id
  }
```